Bug 1298918 - Show Highlights in Activity Stream Panel

sleroux opened the pull request:
First version of implementing a similiar algorithm for determining 'highlights' from a combination of a user's history and bookmarks. The query mostly implements the pseudo query Sebastian wrote here [1] with the except of a host blacklist. Unfortunately we don't store when a bookmark was created on iOS and is not something that comes down from sync bookmarks [2]. The best we can do is use the last_modified time of the bookmark. Probably not the worst thing as in most cases would reflect creation time otherwise it would indicate some source of activity on the bookmark which is significant in itself. One question I have with this approach is that the query assumes local_modified and server_modified come from the same clock which isn't true. Are these comparable in any way?

[1] https://bugzilla.mozilla.org/show_bug.cgi?id=1293710#c5
[2] https://bugzilla.mozilla.org/show_bug.cgi?id=676563

rnewman commented:
> One question I have with this approach is that the query assumes local_modified and server_modified come from the same clock which isn't true. Are these comparable in any way?

They're not comparable, but:

* A user who doesn't use Sync will only have `local_modified`.
* A user who does use Sync, and who is syncing bidirectionally, will eventually have a `server_modified` for every record. See <https://github.com/mozilla/firefox-ios/blob/068ad3cc790247f7d9f2faf1994ec6ca1fe0db43/Storage/SQL/SQLiteBookmarksSyncing.swift#L1039-L1039>
* Usually an iOS user's device will have a somewhat accurate clock, so these will be vaguely comparable.

rnewman commented on Storage/SQL/SQLiteHistoryRecommendations.swift:
```diff
@@ -0,0 +1,97 @@
+import XCGLogger
+import Deferred
+
+/// Small enum to help strength parameter requirements for potential bookmark highlights query below.
```
strengthen

rnewman commented on Storage/SQL/SQLiteHistoryRecommendations.swift:
```diff
@@ -0,0 +1,97 @@
+extension SQLiteHistory: HistoryRecommendations {
+    public func getHighlights() -> Deferred<Maybe<Cursor<Site>>> {
+        let limit = 19
+        let microsecondsPerMinute: Timestamp = 60_000_000 // 1000 * 1000 * 60
```
Technically this isn't a `Timestamp` — [Alan Eliasen would murder you](http://futureboy.us/frinkdata/units.txt). It's a factor.

Just use `UInt64`.

rnewman commented on Storage/SQL/SQLiteHistoryRecommendations.swift:
```diff
@@ -0,0 +1,97 @@
+        let limit = 19
+        let microsecondsPerMinute: Timestamp = 60_000_000 // 1000 * 1000 * 60
+        let thirtyMinutesAgo = NSDate.nowMicroseconds() - 30 * microsecondsPerMinute
+        let threeDaysAgo = NSDate.nowMicroseconds() - (60 * microsecondsPerMinute) * 24 * 3
```
Compute `nowMicroseconds()` once.

rnewman commented on Storage/SQL/SQLiteHistoryRecommendations.swift:
```diff
@@ -0,0 +1,97 @@
+            "   ORDER BY visitDate DESC",
+            "   LIMIT \(historyLimit)",
+            ")",
+            "WHERE visitCount <= 3 AND visitDate < \(thirtyMinutesAgo)",
```
Two problems with this.

Firstly, I think your comparison is backwards — you're asking for timestamps that are numerically less than a moment in time, so this will be *non-recent* history (smaller timestamp).

Secondly, this is a hopelessly inefficient way of phrasing this. Presumably there aren't many sites visited in the last thirty minutes, so make that your most specific clause. Something like:

```
SELECT DISTINCT siteID FROM \(TableVisits) WHERE date > \(thirtyMinutesAgo)
```

is a nice subquery to give you all recently visited sites.

Because you always know the most recent date will be in the candidate rows, you can get the max date for free:

```
SELECT siteID, max(date) AS latest FROM \(TableVisits) WHERE date > \(thirtyMinutesAgo) GROUP BY siteID;
```

You'd need to join back against the visits table to find the count. You use `sum` to do this, but I'd use a subquery in the projection:

```
SELECT s AS siteID, latest, (SELECT COUNT(1) FROM \(TableVisits) WHERE siteID = s) AS visitCount
FROM
(SELECT siteID AS s, max(date) AS latest
 FROM \(TableVisits) WHERE date > \(thirtyMinutesAgo)
 GROUP BY siteID
 LIMIT \(historyLimit))
WHERE visitCount <= 3;
```

rnewman commented on Storage/SQL/SQLiteHistoryRecommendations.swift:
```diff
@@ -0,0 +1,97 @@
+        let bookmarkHighlights = [
+            "SELECT historyID, url, title, guid, visitCount, visitDate",
+            "FROM (",
+            "   SELECT *",
```
Check whether you need this wrapper select. But again, you want to push that modified check and visit count as deep into those queries as you can — you shouldn't need them outside of the `UNION`.

Note that there's already a view that overlays `Local` on top of `Mirror`, so use that — it takes care of handling shadowing. Doing it by hand this way means that if a user locally deletes a highlight bookmark, you'll still show it because it's in the mirror!

sleroux replied:
> Note that there's already a view that overlays Local on top of Mirror, so use that

I'd like to use the view but it has no notion of 'modified' time. I suppose I could add it to the projection but then it would combine `server_modified` and `local_modified` under the same column. Given that the clocks aren't exact, I'm not sure if that would be safe or not.

Thoughts?

rnewman replied:
You could do something like this:

```
    private let localBookmarksView =
    "CREATE VIEW \(ViewBookmarksLocalOnMirror) AS " +
    "SELECT -1 AS id, guid, type, is_deleted, parentid, parentName, feedUri, siteUri, pos, title, description, bmkUri, folderName, faviconID, NULL AS local_modified, server_modified, 0 AS is_overridden " +
    "FROM \(TableBookmarksMirror) WHERE is_overridden IS NOT 1 " +
    "UNION ALL " +
    "SELECT -1 AS id, guid, type, is_deleted, parentid, parentName, feedUri, siteUri, pos, title, description, bmkUri, folderName, faviconID, local_modified, NULL AS server_modified, 1 AS is_overridden " +
    "FROM \(TableBookmarksLocal) WHERE is_deleted IS NOT 1"
```

or combine the two into a single `modified` column and and check `is_overridden` at point of use if you want to know whether it's a local or server time, and don't ever want both the local modification and server modification time for a locally modified bookmark.

rnewman commented:
First-pass feedback done, @sleroux. Lemme know when you want me to look again!

sleroux commented:
@rnewman Thanks for the suggestion to the view on local bookmarks. With the changes to the query and the adjusted view, this looks a lot simpler and seems to be a much faster query. I'll follow up with some measureBlock tests on the query. When you get a chance, could you give this another pass?

farhanpatel commented:
Are you planning to add support for Hiding items from Highlights? Similar to how TopSites work.

sleroux commented:
> Are you planning to add support for Hiding items from Highlights? Similar to how TopSites work.

Good question. I'm not actually sure. Also makes me think - does highlights also include what we want to show for Top Sites or are we keeping that algorithm the same?

cc @bryanbell

bkmunar commented:
@sleroux We do want to support hiding items for highlights per Bryan Bell's request to add a "dismiss" option for the context menu. Also, idk to me I feel like Top Sites algo is the same b/c my desktop top sites seem to be my phone top sites

rnewman commented on Storage/SQL/SQLiteHistoryRecommendations.swift:
```diff
@@ -0,0 +1,64 @@
+            "   FROM (",
+            "       SELECT siteID AS s, max(date) AS visitDate",
+            "       FROM \(TableVisits)",
+            "       WHERE date < \(thirtyMinutesAgo)",
```
`thirtyMinutesAgo` is the only dynamic part of this query. If you extract that as a value in `args`, the compiler should be able to avoid dynamically concatenating this string. (Same below for `threeDaysAgo`.)

In the future, if we need to squeeze more perf out of the system, that will also allow us to pre-parse statements rather than parsing SQL each time we fetch.

rnewman commented on Storage/SQL/SQLiteHistoryRecommendations.swift:
```diff
@@ -0,0 +1,64 @@
+            "   WHERE visitCount <= 3 AND title NOT NULL AND title != ''",
+            "   LIMIT \(historyLimit)",
+            ")"
+        ].joinWithSeparator(" ")
```
Implied: do this with string concatenation instead of dynamically using `joinWithSeparator`. I think the Swift compiler is much better at type analysis now, so this shouldn't blow up compile time, but probably not smart enough to recognize `joinWithSeparator` on constants.

rnewman commented on Storage/SQL/SQLiteHistoryRecommendations.swift:
```diff
@@ -0,0 +1,64 @@
+            "   FROM (",
+            "       SELECT bmkUri",
+            "       FROM \(ViewBookmarksLocalOnMirror)",
+            "       WHERE \(ViewBookmarksLocalOnMirror).server_modified > \(threeDaysAgo) OR \(ViewBookmarksLocalOnMirror).local_modified > \(threeDaysAgo)",
```
Bear in mind that you'll need to enable bidirectional syncing in order to test this feature.

rnewman commented on Storage/SQL/SQLiteHistoryRecommendations.swift:
```diff
@@ -0,0 +1,64 @@
+            "SELECT \(highlightProjection)",
+            "FROM (",
+                bookmarkHighlights,
+                "UNION",
```
This `UNION` might not do what you expect.

Your bookmark highlights always return `NULL AS visitDate`. Your non-recent history query always returns a max visit date. `visitDate` is in `highlightProjection`, so this `UNION` is always equivalent to `UNION ALL`, and you're relying on `GROUP BY url` below to _randomly_ select one of the two rows in the event that you have a bookmark that's been visited more than three times more than 30 minutes ago.

The observed result is probably that sometimes you'll have a date in a row, and sometimes you won't.

sleroux commented:
@rnewman I've rebased from master to support Xcode 8/Swift 2.3 and address the UNION issue. I think by using `max(visitDate)` in with the `GROUP BY` statement we'll always select the one with the date if there is a bookmark as well. I've also added a `isBookmarked` flag to more explicitly indicate that the visit is indeed a bookmark instead of relying on a NULL visitDate for that info. 

I haven't tested this yet with remote bookmarks since I'm seeing an issue on iOS 10 with sync but I'll look into that this afternoon.

rnewman commented:
Figured I'd flesh out my review comment for posterity.

In order to use `UNION` (or even `DISTINCT`!) to get unique results, you need to either narrow the set of columns so that `UNION` removes the duplicates (which means discarding all of the stuff that makes bookmarks and history differ!), use an enclosing query to transform the result set, or…

Conceptually you're trying to do a `FULL OUTER JOIN`: that is, give me things from this set, and things from this set, and if they overlap on a given criterion, merge them together.

SQLite doesn't support `FOJ`, and [emulating it isn't cheap](https://en.wikipedia.org/wiki/Join_(SQL)#Full%5Fouter%5Fjoin), but you could do it.

The approach you've taken here is an enclosing query, using `GROUP BY` to collapse rows together when they refer to the same URL. That needs a teensy bit more work, because the behavior when you return a non-aggregate value from a grouped result set is undefined. Comment inline…

rnewman commented on Storage/SQL/SQLiteHistoryRecommendations.swift:
```diff
@@ -0,0 +1,56 @@
+            ")"
+
+        let highlightsQuery =
+            "SELECT historyID, url, title, guid, visitCount, max(visitDate) AS visitDate, isBookmarked, iconID, iconURL, iconType, iconDate, iconWidth " +
```
Some of these values come from individual rows, and differ on the row. `url` is fixed by virtue of being the grouping characteristic, but `visitCount` might come from the bookmark row, or `isBookmarked` might come from the history row!

Here you see why this is really a `FULL OUTER JOIN`: we want to merge the two together.

Now, we know there's at most one history row, and at most one bookmark URL, for each URL. If a bookmark row is returned, it'll already be joined against history, so the title etc. will match. The remaining column is `isBookmarked`, so we can try to phrase this projection in terms of aggregates:

```
SELECT historyID, url, title, guid, visitCount, max(visitDate) AS visitDate, max(isBookmarked) AS isBookmarked, iconID, iconURL, iconType, iconDate, iconWidth
```

However, we're half-way to just doing a FOJ — you're selecting bookmarks and `LEFT JOIN`ing them against history. Another option here, then, is for you to alter the history side of the `UNION` to be history items that match but _are not bookmarked_. You can then safely avoid grouping by anything, and just use a `UNION ALL`.

sleroux replied:
> Another option here, then, is for you to alter the history side of the UNION to be history items that match but are not bookmarked. You can then safely avoid grouping by anything, and just use a UNION ALL.

That's smart. I feel like I can avoid a lot of hassle doing that instead since I know that the bookmark/recent history queries will be mutually exclusive.

I think I can modify the query to include a bookmark check in the project like so (excludes the date checks):

```
	SELECT historyID, url, title, latest, visitCount FROM (	
		SELECT history.id as historyID, url, title, latest, 
			(SELECT COUNT(1) FROM visits WHERE s = visits.siteID) AS visitCount,
			(SELECT COUNT(1) FROM view_bookmarksLocal_on_mirror WHERE view_bookmarksLocal_on_mirror.bmkUri == url) AS bookmarked
		FROM (
			SELECT siteID AS s, max(date) AS latest
			FROM visits
			GROUP BY siteID
		)
		LEFT JOIN history ON history.id = s
		WHERE visitCount <= 3 AND title NOT NULL AND title != '' AND bookmarked == 0
		LIMIT 19
	)
```
